orttraining/orttraining/python/training/ortmodule/_training_manager.py

```diff
@@ -57,175 +58,193 @@ def forward(self, *inputs, **kwargs):
-                            backward_outputs[initializer_index]))
-                        initializer_index += 1
-                    else:
-                        results.append(None)
-
-                return tuple(results)
-
-        return _io.unflatten_user_output(self._module_output_schema,
-                                        _ORTModuleFunction.apply(
-                                            *_io._combine_input_buffers_initializers(
-                                                self._graph_initializers,
-                                                self._graph_info.user_input_names,
-                                                self._input_info,
-                                                self._flattened_module.named_buffers(),
-                                                inputs,
-                                                kwargs,
-                                                self._device)))
+                            # input_name is not found in the self._input_info.require_grad_names list
+                            # Append None to results for each input that did not require grad
+                            results.append(None)
+
+                    # Append gradients of initializer to results
+                    # Go over each initializer, check if it required grad and append to results accordingly
+                    initializer_index = num_user_input_grads
+                    for initializer_name in self._graph_info.initializer_names:
+                        if initializer_name in self._graph_initializer_names_to_train:
+                            results.append(_utils._torch_tensor_from_dl_pack(
+                                backward_outputs.dlpack_at(initializer_index),
+                                backward_outputs[initializer_index]))
+                            initializer_index += 1
+                        else:
+                            results.append(None)
+
+                    return tuple(results)
+
+            return _io.unflatten_user_output(self._module_output_schema,
+                                             _ORTModuleFunction.apply(
+                                                 *_io._combine_input_buffers_initializers(
+                                                     self._graph_initializers,
+                                                     self._graph_info.user_input_names,
+                                                     self._input_info,
+                                                     self._flattened_module.named_buffers(),
+                                                     inputs,
+                                                     kwargs,
+                                                     self._device)))
+        except ORTModuleFallbackException as e:
+            # Exceptions subject to fallback are handled here
+            self._fallback_manager.handle_exception(e)
```

Comment:
Do you think we could do an auto fallback? For example have an optional arguent in handle_fallback that is the fallback_routine (function partial of ```model(*inputs, **kwargs)```). If this argument is provided, that means handle_exception can auto fallback to the fallback routine. Else, It will return None.

That way, we don't need a three step process to first handle exception and then check if ```is_pending``` and then fallback to pytorch. What are your thoughts?

Reply:
As a design principle, we should avoid having one method doing too many things. Soon we may need to do additional tasks before falling back (such as free memory on the ort backend, etc)

Having a unified API would forces to do everything inside that method or worse, passing a function as parameter to run after deciding to fallback and actually falling back